Add MGCoarseGridApplyPreconditioner

peterrum opened the pull request:
similar to `MGCoarseGridApplySmoother`

kronbichler commented on include/deal.II/multigrid/mg_coarse.h:
```diff
@@ -284,6 +333,99 @@ MGCoarseGridApplySmoother<VectorType>::operator()(const unsigned int level,
+    template <class VectorType,
+              class PreconditionerType,
+              typename std::enable_if<
+                std::is_same<typename VectorType::value_type, double>::value,
+                VectorType>::type * = nullptr>
```
I think it would be nicer if we could introspect whether the `PrecondditionerType` supports `vmult` with `VectorType`. I anticipate possible problems if we add float-based preconditioner classes.

kronbichler commented on include/deal.II/multigrid/mg_coarse.h:
```diff
@@ -284,6 +333,99 @@ MGCoarseGridApplySmoother<VectorType>::operator()(const unsigned int level,
+          VectorType &             dst,
+          const VectorType &       src)
+    {
+      LinearAlgebra::distributed::Vector<double> src_;
+      LinearAlgebra::distributed::Vector<double> dst_;
+
+      src_ = src;
+      dst_ = dst;
+
+      // to allow the case that the preconditioner was only set up on a
+      // subset of processes
+      if (preconditioner != nullptr)
+        preconditioner->vmult(dst_, src_);
```
I think this hard-codes the code to `LA:d:Vector`, which will not work with the PETSc-based preconditioners at the moment. I guess it is a bit out-of-scope right now, but I think we should work towards suitable interfaces also for that latter case. Having written https://github.com/exadg/exadg/blob/b69336a4496096f57b18d32deefbffcfe9028f5f/include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h#L235-L253 and related functionality in an external code, I think something like that in deal.II would be great (but not for this PR, clearly).

kronbichler reviewed (commented):
This looks like a good direction. Before we go on, I would like to ask about opinions on the name. It is not really just a `Preconditioner` we support here, but any **Operator** that provides a `vmult` operation. Hence, I wonder whether the class name `MGCoarseGridApplyOperator` would be more suitable.

bangerth commented:
We haven't made progress with this over the past weeks. I'm going to move it to the next release unless someone wants to step up and finish this today :-)

peterrum commented:
@kronbichler I have finally renamed the class to `MGCoarseGridApplyOperator`. I also removed the additional logic related to float/double (and separate communicators). I think, for now, it is to have the logic in user code. Furthermore, I have added a test.

kronbichler reviewed (approved):
Thank you. This is fine. I'm not quite sure if we have the best name in terms of the class name `...Operator` versus the actual operator being called `MatrixType` and `matrix`, but I see the possible confusion with the key word `operator`, so I do not see a better name and suggest to merge this now. We can always improve things later if we have better ideas.
